client/ipa-getkeytab.c

```diff
@@ -135,61 +158,86 @@ int filter_keys(krb5_context krbctx, struct keys_container *keys,
     return n;
 }
 
-static int ipa_ldap_init(LDAP ** ld, const char * scheme, const char * servername, const int  port)
+static int ipa_server_to_uri(const char *servername, const char *mech,
+                             char **ldap_uri)
 {
-	char* url = NULL;
-	int  url_len = snprintf(url,0,"%s://%s:%d",scheme,servername,port) +1;
+    char *url = NULL;
+    int url_len = 0;
+    int port = 389;
+    char *scheme = "ldap";
 
-	url = (char *)malloc (url_len);
-	if (!url){
-		fprintf(stderr, _("Out of memory \n"));
-		return LDAP_NO_MEMORY;
-	}
-	sprintf(url,"%s://%s:%d",scheme,servername,port);
-	int rc = ldap_initialize(ld, url);
+    url_len = asprintf(&url,"%s://%s:%d",scheme,servername,port);
 
-	free(url);
-	return rc;
+    if (url_len == -1){
+        fprintf(stderr, _("Out of memory \n"));
+        return LDAP_NO_MEMORY;
+    }
+    *ldap_uri = url;
+    return 0;
 }
 
-const char *ca_cert_file = "/etc/ipa/ca.crt";
-
-static int ipa_ldap_bind(const char *server_name, krb5_principal bind_princ,
-			 const char *bind_dn, const char *bind_pw, LDAP **_ld)
+static int ipa_ldap_init(LDAP **ld, const char *ldap_uri)
 {
-    char *msg = NULL;
-    struct berval bv;
-    int version;
-    LDAP *ld;
-    int ssl;
-    int ret;
+    int rc = 0;
+    rc = ldap_initialize(ld, ldap_uri);
 
-    /* TODO: support referrals ? */
-    if (bind_dn) {
-        ret = ldap_set_option(NULL, LDAP_OPT_X_TLS_CACERTFILE, ca_cert_file);
+    return rc;
+}
+
+static int ipa_tls_ssl_init(LDAP *ld, const char *ldap_uri)
+{
+    int ret = LDAP_SUCCESS;
+    int tls_hard = LDAP_OPT_X_TLS_HARD;
+    int tls_demand = LDAP_OPT_X_TLS_DEMAND;
+    char *ldap = "ldap://";
+    char *ldaps = "ldaps://";
```

Comment:
there is no need to use an actual variable, beside these would have to be defined as const char as the strings are constant.

Use defines:
```#define SCHEMA_LDAP "ldap://"```
```#define SCHEMA_LDAPS "ldaps://"```
Then for lenght you can use sizeof(SCHEMA_LDAP) -1